internal/mode/static/telemetry/collector_test.go

```diff
@@ -363,4 +404,136 @@ var _ = Describe("Collector", Ordered, func() {
+									Name:            "pod1",
+									OwnerReferences: nil,
+								}
+							default:
+								Fail(fmt.Sprintf("unknown type: %T", typedObj))
+							}
+							return nil
+						})
+
+					_, err := dataCollector.Collect(ctx)
+					Expect(err).To(HaveOccurred())
+				})
+
+				It("should error if the Pod has multiple owner references", func() {
+					k8sClientReader.GetCalls(
+						func(ctx context.Context, key client.ObjectKey, object client.Object, option ...client.GetOption) error {
+							Expect(option).To(BeEmpty())
+
+							switch typedObj := object.(type) {
+							case *v1.Pod:
+								typedObj.ObjectMeta = metav1.ObjectMeta{
+									Name: "pod1",
+									OwnerReferences: []metav1.OwnerReference{
+										{
+											Kind: "ReplicaSet",
+											Name: "replicaset1",
+										},
+										{
+											Kind: "ReplicaSet",
+											Name: "replicaset2",
+										},
+									},
+								}
+							default:
+								Fail(fmt.Sprintf("unknown type: %T", typedObj))
+							}
+							return nil
+						})
+
+					_, err := dataCollector.Collect(ctx)
+					Expect(err).To(HaveOccurred())
+				})
+
+				It("should error if the Pod's owner reference is not a ReplicaSet", func() {
+					k8sClientReader.GetCalls(
+						func(ctx context.Context, key client.ObjectKey, object client.Object, option ...client.GetOption) error {
+							Expect(option).To(BeEmpty())
+
+							switch typedObj := object.(type) {
+							case *v1.Pod:
+								typedObj.ObjectMeta = metav1.ObjectMeta{
+									Name: "pod1",
+									OwnerReferences: []metav1.OwnerReference{
+										{
+											Kind: "Deployment",
+											Name: "deployment1",
+										},
+									},
+								}
+							default:
+								Fail(fmt.Sprintf("unknown type: %T", typedObj))
+							}
+							return nil
+						})
+
+					_, err := dataCollector.Collect(ctx)
+					Expect(err).To(HaveOccurred())
+				})
+
+				It("should error if the replica set's replicas is nil", func() {
+					k8sClientReader.GetCalls(
+						func(ctx context.Context, key client.ObjectKey, object client.Object, option ...client.GetOption) error {
+							Expect(option).To(BeEmpty())
+
+							switch typedObj := object.(type) {
+							case *v1.Pod:
+								typedObj.ObjectMeta = metav1.ObjectMeta{
+									Name: "pod1",
+									OwnerReferences: []metav1.OwnerReference{
+										{
+											Kind: "ReplicaSet",
+											Name: "replicaset1",
+										},
+									},
+								}
+							case *appsv1.ReplicaSet:
+								typedObj.Spec = appsv1.ReplicaSetSpec{
+									Replicas: nil,
+								}
+							default:
+								Fail(fmt.Sprintf("unknown type: %T", typedObj))
+							}
+							return nil
+						})
+
+					_, err := dataCollector.Collect(ctx)
+					Expect(err).To(HaveOccurred())
+				})
+
+				It("should error if the kubernetes client errored when getting the ReplicaSet", func() {
+					k8sClientReader.GetReturnsOnCall(1, errors.New("there was an error"))
```

Comment:
The approach with specifying the order here is a bit brittle. 
What if we add some other collector that also call Gets and change the order of get calls?

One away to make it less brittle is include logic in Get fake so that it returns errors if the resource doesn't exist (meaning the fake client checks the passed namespace and name, and if it doesn't have the resource, it returns an error). Or the fake client return errors for resources with special names.

Reply:
I've changed this part of code to a `k8sClientReader.GetCalls...` which switches on the type of object and, in this case, returns an error if its of type ReplicaSet. 

I'm also a little concerned with how testing will look if we introduce many other calls to Get/List, however I think this current implementation of specifying the `GetCalls` can work as we get to switch on the Object type and can use the `client.ObjectKey` to give us more room with testing a specific object. (maybe in the future we split up the control / data plane and have two pods, each that can scale and when getting the Pod we can use the objectkey (the name) to specify which Pod we want). Thoughts on how it looks now?